File: Homework/Hang_Nguyen/lesson_1/subsets.py
class Solution:
    def get_bit(self, num, bit):
        return 0 if num & (1 << bit) == 0 else 1

    def subsets(self, nums):
        subsets_size = pow(2, len(nums))
        subsets = []
        for i in range(0, subsets_size):
            subs = []
            for j in range(0, len(nums)):
                # print(str(i) + "------" + str(j) + "------" + str(self.get_bit(i, j)))
                if self.get_bit(i, j) == 1:
                    subs.append(nums[j])
            # print("---------------------------------------------------------")
            subsets.append(subs)
        return subsets


# TEST CASE
solution = Solution()

nums = [1,2,3]
print("Subsets for TEST 1 is: " + str(solution.subsets(nums)))

nums = [0]
print("Subsets for TEST 2 is: " + str(solution.subsets(nums)))

File: Homework/Le_Trong/src/lesson_01/subsets.rs
// Integer[] -> Integer[][]
// produce all possible subsets of the given list of integers.

/**
 * Runtime Complexity: O(2^n * n) với n là nums.len()
 * Space Complexity: O(2^n * n) với n là nums.len()
 */
pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut result: Vec<Vec<i32>> = vec![];

    for i in 0..2_i32.pow(nums.len() as u32) {
        let mut subset: Vec<i32> = vec![];
        for j in 0..nums.len() {
            let bit = (i >> j) & 1;
            if bit == 1 {
                subset.push(nums[j]);
            }
        }
        result.push(subset);
    }

    return result;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_subsets() {
        assert_eq!(
            subsets(vec![1, 2, 3]),
            vec![
                vec![],
                vec![1],
                vec![2],
                vec![1, 2],
                vec![3],
                vec![1, 3],
                vec![2, 3],
                vec![1, 2, 3],
            ]
        );
    }
}
